parsers/NED.py

```diff
@@ -145,6 +172,78 @@ def call_api(target_datetime: datetime, forecast: bool = False):
     return results
 
 
+def _get_entsoe_production_data(
+    zone_key: ZoneKey,
+    session: Session,
+    target_datetime: datetime,
+    logger: Logger,
+) -> ProductionBreakdownList:
+    # Add 2 days as ENTSOE fetches data from three days before target_datetime, where NED fetches for target_datetime and day before
+    ENTSOE_raw_data = ENTSOE_query_production(
+        ENTSOE_DOMAIN_MAPPINGS[zone_key],
+        session,
+        target_datetime=(target_datetime + timedelta(days=2)),
+    )
+    if ENTSOE_raw_data is None:
+        raise ParserException(
+            parser="NED.py",
+            message="Failed to fetch ENTSOE data",
+            zone_key=zone_key,
+        )
+    ENTSOE_parsed_data = ENTSOE_parse_production(
+        ENTSOE_raw_data, zoneKey=zone_key, logger=logger
+    )
+
+    return ENTSOE_parsed_data
+
+
+def combine_production_data(entsoe_data, ned_data, logger):
+    non_matching_indices = set()
+
+    # Allocate the productions from ENTSOE as NED.nl is missing all types except wind and solar 2017-2021
+    # Before 2017, NED.nl is missing all types except wind
+    for idx, ned_event in enumerate(ned_data.events):
+        if ned_event.datetime >= datetime(2021, 1, 1, tzinfo=timezone.utc):
+            continue  # Skip events after 2021, as NED has all types
+
+        entsoe_events = [
+            e for e in entsoe_data.events if e.datetime == ned_event.datetime
+        ]
+
+        if len(entsoe_events) == 0:
+            non_matching_indices.add(idx)
+            continue
+
+        entsoe_event = entsoe_events[0]
+
+        # NED solar and wind data available for 2017-2021 span
+        for key in UNAVAILABLE_TYPES_BEFORE2021:
+            attr = getattr(entsoe_event.production, key, None)
+            if attr is not None:
+                setattr(ned_event.production, key, attr)
+
+        # Only NED wind data available before 2017
+        if ned_event.datetime < datetime(2017, 1, 1, tzinfo=timezone.utc):
+            for key in ADDITIONAL_UNAVAILABLE_TYPES_BEFORE2017:
+                attr = getattr(entsoe_event.production, key, None)
+                if attr is not None:
+                    setattr(ned_event.production, key, attr)
+
+        ned_event.source += "," + entsoe_event.source
+
+    if len(non_matching_indices) > 0:
+        logger.info(
+            f"Failed to match {len(non_matching_indices)}  NED events with ENTSOE events"
+        )
+        ned_data.events = [
+            event
+            for idx, event in enumerate(ned_data.events)
+            if idx not in non_matching_indices
+        ]
+
+    return ned_data
```

Comment:
We have this function on the ProductionBreakdownList class:
https://github.com/electricitymaps/electricitymaps-contrib/blob/64987b5bafdb2fff0be3fafa9e00b5219f000e01/electricitymap/contrib/lib/models/event_lists.py#L263-L304

Would that work instead of this custom one?
If so I think it would be better to use that so it is consistent with all other parsers.

Reply:
Ah okay, I will look into using that instead. I had worked on an outdated branch, so I closed this commit as I'm not very confident in working in git yet:)

Reply:
That's okay! I don't mind at all 🙂

Reply:
@VIKTORVAV99 Hmm, I think it might be a bit challenging to use this function. 

Example if we are fetching data for a span 2016-12-31 to 2017-01-01, after 2017 we get wind and solar from NED and the rest from entsoe, but before 2017 we only get wind from NED and the rest from entsoe. So in order to use this function we would have to set the entsoe wind and solar values to zero after 2017, and only wind to zero before 2017, so we don't double up on wind or solar when merging the two frames. 

Perhaps this is still better in order to keep it consistent? What do you think? :)

Reply:
I see!

We also have a similar static function that just updates values instead of merging them, perhaps that's better then?

If so we can fetch the ENTSOE data then the NED NL data and overwrite all ENTSOE data with the data that is available from NED NL.

That sounds like it should do what we want?

But if that don't work either I'm fine with keeping the current function if we add tests for it so we don't accidentally change something in the parser classes that break this function.

Reply:
I think that would work! 
I assume that it is the update_production_breakdowns function from the ProductionBreakdownList class? I will look into it, thanks!

Reply:
Yeah that should be the one!